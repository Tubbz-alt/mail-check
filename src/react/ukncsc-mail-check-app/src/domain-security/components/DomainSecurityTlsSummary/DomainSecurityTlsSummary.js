import React from 'react';
import PropTypes from 'prop-types';
import { Link } from 'react-router-dom';
import map from 'lodash/map';
import { Header, Message } from 'semantic-ui-react';
import {
  DomainSecurityMessage,
  DomainSecuritySummaryMessages,
} from 'domain-security/components';
import { DomainSecurityContext } from 'domain-security/context';

const DomainSecurityTlsSummary = ({
  description,
  id,
  loading,
  error,
  records,
  pending,
}) => (
  <React.Fragment>
    <p>{description}</p>
    <p>
      <Link to="/domain-security/tls-advice">View NCSC advice on TLS</Link>
    </p>
    {!loading &&
      !error && (
        <React.Fragment>
          {map(records, record => (
            <React.Fragment key={record.id}>
              <Header as="h2">{record.hostname}</Header>
              <p>
                <DomainSecurityContext.Consumer>
                  {value => (
                    <Link to={`/${value}/${id}/tls/${record.id}`}>
                      View more information
                    </Link>
                  )}
                </DomainSecurityContext.Consumer>
              </p>
              <DomainSecuritySummaryMessages
                id={record.id}
                type="TLS"
                failures={record.failures.length}
                warnings={record.warnings.length}
                inconclusives={record.inconclusives.length}
              />
            </React.Fragment>
          ))}
          {!pending &&
            !records.length && (
              <DomainSecurityMessage success>
                No MX records found.
              </DomainSecurityMessage>
            )}
          {pending && (
            <DomainSecurityMessage info={pending}>
              No MX records found yet.
            </DomainSecurityMessage>
          )}
        </React.Fragment>
      )}
    {error && (
      <Message error>
        There was a problem retrieving the TLS summary for this domain:{' '}
        {error.message}
      </Message>
    )}
  </React.Fragment>
);

DomainSecurityTlsSummary.defaultProps = {
  id: null,
  loading: false,
  error: null,
  records: [],
  pending: false,
};

DomainSecurityTlsSummary.propTypes = {
  description: PropTypes.string.isRequired,
  id: PropTypes.string,
  loading: PropTypes.bool,
  error: PropTypes.shape({ message: PropTypes.string }),
  records: PropTypes.arrayOf(
    PropTypes.shape({
      id: PropTypes.string,
      hostname: PropTypes.string,
      failures: PropTypes.arrayOf(PropTypes.string),
      warnings: PropTypes.arrayOf(PropTypes.string),
      inconclusives: PropTypes.arrayOf(PropTypes.string),
    })
  ),
  pending: PropTypes.bool,
};

export default DomainSecurityTlsSummary;
