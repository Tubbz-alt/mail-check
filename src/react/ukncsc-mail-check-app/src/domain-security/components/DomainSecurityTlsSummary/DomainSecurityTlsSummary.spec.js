import React from 'react';
import { BrowserRouter } from 'react-router-dom';
import { render } from 'react-testing-library';
import 'dom-testing-library/extend-expect';
import DomainSecurityTlsSummary from './DomainSecurityTlsSummary';

describe('DomainSecurityTlsSummary', () => {
  let container;

  describe('when no records are provided', () => {
    beforeEach(() => {
      ({ container } = render(
        <BrowserRouter>
          <DomainSecurityTlsSummary records={[]}>
            <p>foo</p>
          </DomainSecurityTlsSummary>
        </BrowserRouter>
      ));
    });

    test('it should match the snapshot', () =>
      expect(container).toMatchSnapshot());

    test('it should display a success message', () =>
      expect(
        container.getElementsByClassName('ui success message')[0]
      ).toHaveTextContent('No MX records found.'));
  });

  describe('when no failures, warnings or inconclusives are present', () => {
    beforeEach(() => {
      ({ container } = render(
        <BrowserRouter>
          <DomainSecurityTlsSummary
            records={[
              {
                id: '1',
                hostname: '1.com',
                failures: [],
                warnings: [],
                inconclusives: [],
              },
            ]}
          >
            <p>bar</p>
          </DomainSecurityTlsSummary>
        </BrowserRouter>
      ));
    });

    test('it should display a success message', () =>
      expect(
        container.getElementsByClassName('ui success message')[0]
      ).toHaveTextContent('TLS is well configured.'));
  });

  describe('when an error is present', () => {
    beforeEach(() => {
      ({ container } = render(
        <BrowserRouter>
          <DomainSecurityTlsSummary error={Error('oh noes!')}>
            <p>flum</p>
          </DomainSecurityTlsSummary>
        </BrowserRouter>
      ));
    });

    test('it should display a message with the error', () =>
      expect(
        container.getElementsByClassName('ui error message')[0]
      ).toHaveTextContent('oh noes!'));
  });
});
