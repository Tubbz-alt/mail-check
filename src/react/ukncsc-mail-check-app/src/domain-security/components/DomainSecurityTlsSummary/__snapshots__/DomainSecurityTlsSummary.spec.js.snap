// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DomainSecurityTlsSummary when no records are provided it should match the snapshot 1`] = `
<div>
  <p>
    foo
  </p>
  <p>
    <a
      href="/domain-security/tls-advice"
    >
      View NCSC advice on TLS
    </a>
  </p>
  <div
    class="ui large success message DomainSecurityMessage--container"
  >
    No MX records found.
  </div>
</div>
`;
